pkg/dataplane/server/dataplane.go

```diff
@@ -40,6 +41,8 @@ type Dataplane struct {
 	clusters           map[string]*cluster.Cluster
 	listeners          map[string]*listener.Listener
 	listenerEnd        map[string]chan bool
+	clustersLock       sync.Mutex
```

Comment:
I think you're missing taking a lock on `GetClusterTarget` and `GetClusterHost`.

Reply:
Done